### ISSUE TITLE
Potential TOCTTOU issue

### ISSUE THREAD

**Issue by toonn** [haskell/cabal#7150]
I'm looking at the code related to reading `cabal.project` files and I noticed some checks for file existence, [like here](https://github.com/haskell/cabal/blob/master/cabal-install/src/Distribution/Client/ProjectConfig.hs#L515).
These seem like [TOCTTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issues to me. Is there a reason not to rely on a "file not found error" or similar to fall back to the case of the file not existing?
I haven't looked into `monitorFiles` so my impression could very well be wrong.

**Comment by hgolden** [comment]
I am just learning about TOCTTOU from the linked Wikipedia page, so maybe I'm missing your point, but it seems to me that the example in the Wikipedia page shows both writing (not reading) a file and also using setuid to access an otherwise inaccessible file. Does cabal do setuid or ever run with root privileges? If not, I don't see the risk in this case.

**Comment by toonn** [comment]
That's just a specific example, in general this is unrelated to setuid. The existence of the file is checked and depending on its existence it is read or monitored. But by the time cabal tries to read or monitor the file the existence may already have changed. The pattern I usually see is to read and catch the error if the file doesn't exist then fall back to whatever you would do if the file doesn't exist.

**Comment by hgolden** [comment]
I agree with the pattern you mention (try to read and fail if the file isn't there). Thank you for your prompt reply.
